common/src/math.rs

```diff
@@ -53,79 +126,268 @@ impl<N: Scalar> Deref for MIsometry<N> {
+    /// Casts the components to an `f32`
     #[inline]
-    pub fn unit_quaternion_to_homogeneous(rotation: na::UnitQuaternion<N>) -> Self {
-        MIsometry(rotation.to_homogeneous())
+    pub fn to_f32(self) -> MUnitPointVector<f32> {
+        MUnitPointVector(self.0.to_f32())
     }
+}
+
+impl MUnitDirectionVector<f32> {
+    /// Casts the components to an `f64`
     #[inline]
-    pub fn rotation_to_homogeneous(rotation: na::Rotation3<N>) -> Self {
-        MIsometry(rotation.to_homogeneous())
+    pub fn to_f64(self) -> MUnitDirectionVector<f64> {
+        MUnitDirectionVector(self.0.to_f64())
     }
+}
+
+impl MUnitDirectionVector<f64> {
+    /// Casts the components to an `f32`
     #[inline]
-    pub fn row(self, i: usize) -> na::RowVector4<N> {
-        self.0.row(i).into()
+    pub fn to_f32(self) -> MUnitDirectionVector<f32> {
+        MUnitDirectionVector(self.0.to_f32())
     }
+}
+
+impl<N: RealField + Copy> AsRef<[[N; 4]; 4]> for MIsometry<N> {
+    #[inline]
+    fn as_ref(&self) -> &[[N; 4]; 4] {
+        self.0.as_ref()
+    }
+}
+
+impl<N: RealField + Copy> From<na::UnitQuaternion<N>> for MIsometry<N> {
+    /// Converts a quaternion into the matrix for the rotation it represents.
+    fn from(value: na::UnitQuaternion<N>) -> Self {
+        MIsometry(value.to_homogeneous())
+    }
+}
+
+impl<N: RealField + Copy> From<na::Rotation3<N>> for MIsometry<N> {
+    /// Converts a rotation into the matrix representing that rotation.
+    fn from(value: na::Rotation3<N>) -> Self {
+        MIsometry(value.to_homogeneous())
+    }
+}
+
+impl<N: RealField + Copy> MIsometry<N> {
+    /// Returns a view containing the i-th row of this matrix.
+    #[inline]
+    pub fn row(&self, i: usize) -> na::MatrixView1x4<'_, N, na::U1, na::U4> {
+        self.0.row(i)
+    }
+
+    /// Creates an identity matrix.
     #[inline]
     pub fn identity() -> Self {
         Self(na::Matrix4::identity())
     }
-    #[inline]
-    pub fn map<N2: Scalar, F: FnMut(N) -> N2>(&self, f: F) -> MIsometry<N2> {
-        MIsometry(self.0.map(f))
+
+    /// The reflection about the hyperbolic plane represented by the given
+    /// normalized vector. An incorrect matrix will be returned if `a` and `b`
+    /// are not normalized.
+    pub fn reflection(normal: &MVector<N>) -> Self {
+        // The formula below is the equivalent of the formula for the
+        // Householder matrix, but the minkowski outer product instead of the
+        // standard outer product to ensure that the reflection is done in
+        // Minkowski space. The resulting formula is
+        // `I - 2vv*`
+        Self(
+            na::Matrix4::<N>::identity()
+                - normal.minkowski_outer_product(normal) * na::convert::<_, N>(2.0),
+        )
+    }
+
+    /// The matrix that translates `a` to `b` given that `a` and `b` are
+    /// normalized point-like `MVectors`. An incorrect matrix will be returned
+    /// if `a` and `b` are not normalized.
```

Comment:
We don't need to document the invariants on the unit types every time we use them. That's what the type is for!

Reply:
Whoops! That was mistakenly left in from older code I was working with. Thanks for the reminder!